proof_ipa.go

```diff
@@ -177,26 +233,34 @@ func SerializeProof(proof *Proof) (*VerkleProof, StateDiff, error) {
 			stemdiff = &statediff[len(statediff)-1]
 			copy(stemdiff.Stem[:], key[:31])
 		}
-		var valueLen = len(proof.Values[i])
+		stemdiff.SuffixDiffs = append(stemdiff.SuffixDiffs, SuffixStateDiff{Suffix: key[31]})
+		newsd := &stemdiff.SuffixDiffs[len(stemdiff.SuffixDiffs)-1]
+
+		var valueLen = len(proof.PreValues[i])
+		switch valueLen {
+		case 0:
+			// null value
+		case 32:
+			newsd.CurrentValue = (*[32]byte)(proof.PreValues[i])
+		default:
+			var aligned [32]byte
+			copy(aligned[:valueLen], proof.PreValues[i])
+			newsd.CurrentValue = (*[32]byte)(unsafe.Pointer(&aligned[0]))
+		}
+
+		valueLen = len(proof.PostValues[i])
 		switch valueLen {
 		case 0:
-			stemdiff.SuffixDiffs = append(stemdiff.SuffixDiffs, SuffixStateDiff{
-				Suffix: key[31],
-			})
+			// null value
 		case 32:
-			stemdiff.SuffixDiffs = append(stemdiff.SuffixDiffs, SuffixStateDiff{
-				Suffix:       key[31],
-				CurrentValue: (*[32]byte)(proof.Values[i]),
-			})
+			newsd.NewValue = (*[32]byte)(proof.PostValues[i])
 		default:
 			var aligned [32]byte
-			copy(aligned[:valueLen], proof.Values[i])
-			stemdiff.SuffixDiffs = append(stemdiff.SuffixDiffs, SuffixStateDiff{
-				Suffix:       key[31],
-				CurrentValue: (*[32]byte)(unsafe.Pointer(&aligned[0])),
-			})
+			copy(aligned[:valueLen], proof.PostValues[i])
+			newsd.NewValue = (*[32]byte)(unsafe.Pointer(&aligned[0]))
```

Comment:
To be honest, I'd add a TODO here to remove the `unsafe` usage. 
If we do the change of removing the `*[32]byte` indirection, we can simply do a `copy` here or even directly using the copy one line above.

Reply:
Actually from Go 1.17 we should be able to do `(*[4]byte)(slice)`